[red-knot] Do not assume that `x != 0` if `x` inhabits `~Literal[0]` (#17370)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/type_api.md

    reveal_type(n2)  # revealed: int
    reveal_type(n3)  # revealed: ~int

def static_truthiness(not_one: Not[Literal[1]]) -> None:
    static_assert(not_one != 1)

>     static_assert(not (not_one == 1))

Reviewer: Rather than deleting this test entirely, I think it would be good to replace it with something like this:

```py
def static_truthiness(not_one: Not[Literal[1]]) -> None:
    # these are both boolean-literal types,
    # since all possible runtime objects that are created by the literal syntax `1`
    # are members of the type `Literal[1]`
    reveal_type(not_one is not 1)  # revealed: Literal[True]
    reveal_type(not_one is 1)  # revealed: Literal[False]

    # But these are both `bool`, rather than `Literal[True]` or `Literal[False]`
    # as there are many runtime objects that inhabit the type `~Literal[1]`
    # but still compare equal to `1`. Two examples are `1.0` and `True`.
    reveal_type(not_one != 1)  # revealed: bool
    reveal_type(not_one == 1)  # revealed: bool
```

    reveal_type(n2)  # revealed: int
    reveal_type(n3)  # revealed: ~int

# error: "Special form `knot_extensions.Not` expected exactly one type parameter"
n: Not[int, str]

def static_truthiness(not_one: Not[Literal[1]]) -> None:

>     # these are both boolean-literal types,
>     # since all possible runtime objects that are created by the literal syntax `1`
>     # are members of the type `Literal[1]`
>     reveal_type(not_one is not 1)  # revealed: bool
>     reveal_type(not_one is 1)  # revealed: bool

Reviewer: this comment doesn't seem to reflect the code immediately below it anymore. The comment says "these are both boolean-literal types", but the assertion reveals `bool`, not `Literal[True]` or `Literal[False]`?

`bool` isn't incorrect here, but it would be great if we could fix this comment to something like

```suggestion
    # TODO: these be `Literal[True]` and `Literal[False]` respectively,
    # since all possible runtime objects that are created by the literal syntax `1`
    # are members of the type `Literal[1]`
    reveal_type(not_one is not 1)  # revealed: bool
    reveal_type(not_one is 1)  # revealed: bool
```

Author: https://github.com/astral-sh/ruff/pull/17425
